hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/CarbonitedMetadataThawingActor.scala

```diff
@@ -2,96 +2,148 @@ package cromwell.services.metadata.hybridcarbonite
+import cromwell.services.metadata.{MetadataQuery, MetadataQueryJobKey}
+import cromwell.services.{BuiltMetadataResponse, FailedMetadataResponse}
 import cromwell.util.JsonEditor
 import io.circe.parser._
 import io.circe.{Json, Printer}
+import spray.json._
 
 import scala.concurrent.{ExecutionContext, Future}
 
-final class CarbonitedMetadataThawingActor(carboniterConfig: HybridCarboniteConfig, serviceRegistry: ActorRef, ioActor: ActorRef) extends LoggingFSM[CarbonitedMetadataThawingState, CarbonitedMetadataThawingData] {
+final class CarbonitedMetadataThawingActor(carboniterConfig: HybridCarboniteConfig, serviceRegistry: ActorRef, ioActor: ActorRef) extends LoggingFSM[ThawingState, ThawingData] {
 
   implicit val ec: ExecutionContext = context.dispatcher
 
   val asyncIo = new AsyncIo(ioActor, DefaultIoCommandBuilder)
 
-  startWith(CarbonitedMetadataThawingActorPendingState, CarbonitedMetadataThawingActorPendingData)
+  startWith(PendingState, PendingData)
 
-  when(CarbonitedMetadataThawingActorPendingState) {
-    case Event(ThawCarbonitedMetadata(rootWorkflowId), CarbonitedMetadataThawingActorPendingData) =>
-
-      val response = for {
-        rawResponseFromIoActor <- asyncIo.contentAsStringAsync(carboniterConfig.makePath(rootWorkflowId), maxBytes = Option(30 * 1024 * 1024), failOnOverflow = true)
+  when(PendingState) {
+    case Event(action: WorkflowMetadataReadAction, PendingData) =>
+      val gcsResponse = for {
+        rawResponseFromIoActor <- asyncIo.contentAsStringAsync(carboniterConfig.makePath(action.workflowId), maxBytes = Option(30 * 1024 * 1024), failOnOverflow = true)
         parsedResponse <- Future.fromTry(parse(rawResponseFromIoActor).toTry.map(GcsMetadataResponse.apply))
       } yield parsedResponse
 
-      response pipeTo self
-      serviceRegistry ! GetRootAndSubworkflowLabels(rootWorkflowId)
+      gcsResponse pipeTo self
+
+      val targetState = if (action.requiresLabels) {
+        serviceRegistry ! GetRootAndSubworkflowLabels(action.workflowId)
+        NeedsGcsAndLabelsState
+      } else {
+        NeedsOnlyGcsState
+      }
 
-      goto(CarbonitedMetadataThawingActorRunningState) using CarbonitedMetadataThawingActorDataNothingYet(sender())
+      goto(targetState) using WorkingNoData(sender(), action)
   }
 
-  when(CarbonitedMetadataThawingActorRunningState) {
+  when(NeedsOnlyGcsState) {
+    case Event(GcsMetadataResponse(gcsResponse), WorkingNoData(requester, action)) =>
+      val responseJson = gcsResponse.editFor(action).toSpray
+      requester ! BuiltMetadataResponse(action, responseJson)
+      stop()
+  }
 
+  when(NeedsGcsAndLabelsState) {
     // Successful Responses:
-    case Event(GcsMetadataResponse(response), CarbonitedMetadataThawingActorDataNothingYet(requester)) =>
-      stay() using CarbonitedMetadataThawingActorDataWithGcsResponse(requester, response)
+    case Event(GcsMetadataResponse(gcsResponse), WorkingSomeData(requester, action, _ , Some(labels))) =>
```

Comment:
TOL: I would make the `WorkingSomeData` more ADT and less GDT by splitting it into the two valid `WorkingWithLabelsData` and `WorkingWithJsonData` options (and remove the possibility of creating invalid `WorkingSomeData(..., None, None)` and `WorkingSomeData(..., Some, Some)` variants by accident at the same time)